buildSrc/subprojects/parser/src/main/kotlin/org/apache/calcite/buildtools/parser/DialectGenerateTask.kt

```diff
@@ -0,0 +1,210 @@
+import org.gradle.api.DefaultTask
+import org.gradle.api.model.ObjectFactory
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.InputDirectory
+import org.gradle.api.tasks.TaskAction
+
+open class DialectGenerateTask @Inject constructor(
+    objectFactory: ObjectFactory
+) : DefaultTask() {
+
+    @InputDirectory
+    val dialectDirectory = objectFactory.directoryProperty()
+
+    @InputDirectory
+    val rootDirectory = objectFactory.directoryProperty()
+
+    @Input
+    var outputFile = ""
+
+    @TaskAction
+    fun run() {
+        val functionMap = extractFunctions()
+    }
+
+    /**
+     * Traverses the parsing directory structure and extracts all of the
+     * functions located in *.ftl files into a Map
+     *
+     */
+    private fun extractFunctions(): Map<String, String> {
+        val rootDirectoryFile = rootDirectory.get().asFile
+        val queue = getTraversalPath(rootDirectoryFile)
+        val functionMap: MutableMap<String, String> = HashMap<String, String>()
+        traverse(queue, rootDirectoryFile, functionMap)
+        return functionMap
+    }
+
+    /**
+     * Generates the parserImpls.ftl file for the dialect.
+     *
+     * @param functions The functions to place into the output file
+     */
+    private fun generateParserImpls(functions: Map<String, String>) {}
+
+    /**
+     * Gets the traversal path for the dialect by "subtracting" the root
+     * absolute path from the dialect directory absolute path.
+     *
+     * @param rootDirectoryFile The file for the root parsing directory
+     */
+    private fun getTraversalPath(rootDirectoryFile: File): Queue<String> {
+        val dialectDirectoryFile = dialectDirectory.get().asFile
+        var dialectPath = dialectDirectoryFile.absolutePath
+        val rootPath = rootDirectoryFile.absolutePath
+        val rootIndex = dialectPath.indexOf(rootPath)
+        dialectPath = dialectPath.substring(rootIndex + rootPath.length + 1)
+        return LinkedList(dialectPath.split("/"))
+    }
+
+    /**
+     * Traverses the determined path given by the queue. Once the queue is
+     * empty, the dialect directory has been reached. In that case any *.ftl
+     * file should be processed and no further traversal should happen.
+     *
+     * @param directories The directories to traverse in topdown order
+     * @param currentDirectory The current directory the function is processing
+     * @param functionMap The map to which the parsing functions will be added to
+     */
+    private fun traverse(
+        directories: Queue<String>,
+        currentDirectory: File,
+        functionMap: MutableMap<String, String>
+    ) {
+        val files = currentDirectory.listFiles()
+        // Ensures that files are processed first.
+        files.sortBy { it.isDirectory }
+        val nextDirectory = directories.peek()
+        for (f in files) {
+            if (f.isFile && f.extension == "ftl") {
+                processFile(f, functionMap)
```

Comment:
Should we have a break in the body of this if or, alternatively, should the following if be an else-if?

Reply:
Good point, will change.